app.js

```diff
@@ -130,6 +132,97 @@ if (urlParsingError) {
+
+    const isSrv = cs.protocol && cs.protocol.includes('srv');
+
+    if (!isSrv) {
+      return raw; // Non-SRV URIs are fine as-is
+    }
+
+    // For SRV URIs, resolve the actual hosts and ports
+    // Some versions of mongodb-connection-string-url use placeholder for hostname
+    // but populate hosts array correctly, so prefer hosts[0] over hostname
+    let hostname = cs.hostname;
+    if (!hostname || hostname === '__this_is_a_placeholder__') {
+      if (cs.hosts && cs.hosts.length > 0) {
+        hostname = cs.hosts[0];
+        console.log('Using hostname from hosts array:', hostname);
+      } else {
+        console.log('No valid hostname found, using original connection string');
+        return raw;
+      }
+    }
+
+    try {
+      const srvRecords = await dns.resolveSrv(`_mongodb._tcp.${hostname}`);
+      if (!Array.isArray(srvRecords) || srvRecords.length === 0) {
+        // Fallback to hostname with default port if SRV resolution fails
+        const fallbackHost = `${hostname}:27017`;
+        const hostList = [fallbackHost];
+
+        let auth = '';
+        if (cs.username) {
+          auth += encodeURIComponent(cs.username);
+          if (cs.password) auth += `:${encodeURIComponent(cs.password)}`;
+          auth += '@';
+        }
+
+        const pathname = cs.pathname || '';
+        const params = cs.searchParams?.toString() || '';
+        const query = params ? `?${params}` : '';
+
+        return `mongodb://${auth}${hostList.join(',')}${pathname}${query}`;
+      }
+
+      // Use the resolved SRV records
+      const hostList = srvRecords.map((record) => `${record.name}:${record.port}`);
+
+      let auth = '';
+      if (cs.username) {
+        auth += encodeURIComponent(cs.username);
+        if (cs.password) auth += `:${encodeURIComponent(cs.password)}`;
+        auth += '@';
+      }
+
+      const pathname = cs.pathname || '';
+      const params = cs.searchParams?.toString() || '';
+      const query = params ? `?${params}` : '';
+
+      return `mongodb://${auth}${hostList.join(',')}${pathname}${query}`;
+    } catch (dnsError) {
+      console.warn('Failed to resolve SRV record for', hostname, ':', dnsError.message);
+      return raw; // Fallback to original if DNS resolution fails
+    }
+  } catch (_e) {
+    return raw; // Fallback to original if parsing fails
+  }
+}
+
+// Precompute client-safe connection strings
+async function initializeClientSafeConnectionStrings() {
+  await Promise.all(
+    mongoURIs.map(async (entry) => {
+      entry.clientConnectionString = await createClientSafeConnectionString(entry.raw);
+      console.log(`Converted connection string for ${entry.id}:`);
+      console.log(`  Original: ${entry.raw}`);
+      console.log(`  Client-safe: ${entry.clientConnectionString}`);
+    })
+  );
+}
+
+// Initialize connection strings
+initializeClientSafeConnectionStrings().catch((err) => {
```

Comment:
It would be better to put `initializeClientSafeConnectionStrings` and `fastify.listen` in a chain

```js
initializeClientSafeConnectionStrings().then(() => fastify.listen(...))
```